js/manager.js

```diff
@@ -1,149 +1,144 @@
+/***************************************************************************
+* Copyright (c) 2018, Voila contributors                                   *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+import { RenderMimeRegistry, standardRendererFactories } from '@jupyterlab/rendermime';
+import { WidgetManager as JupyterLabManager } from '@jupyter-widgets/jupyterlab-manager/lib/manager';
+import { HTMLManager, requireLoader } from '@jupyter-widgets/html-manager';
+import { WidgetRenderer } from '@jupyter-widgets/jupyterlab-manager/lib/renderer';
+import * as output from '@jupyter-widgets/jupyterlab-manager/lib/output';
+import * as base from '@jupyter-widgets/base';
+import * as controls from '@jupyter-widgets/controls';
+
+if (typeof window !== "undefined" && typeof window.define !== "undefined") {
+    window.define("@jupyter-widgets/base", base);
+    window.define("@jupyter-widgets/controls", controls);
+    window.define("@jupyter-widgets/output", output);
+  }
+
+const WIDGET_MIMETYPE = 'application/vnd.jupyter.widget-view+json';
+const htmlManager = new HTMLManager({ loader: requireLoader });
+
+export class WidgetManager extends JupyterLabManager {
+
+    constructor(kernel) {
+        const context = createContext(kernel);
+        const rendermime = createRenderMimeRegistry();
+        super(context, rendermime);
+        this._registerWidgets();
+        this.loader = requireLoader;
+    }
+
+    async build_widgets() {
+        const models = await this._build_models();
+        const tags = document.body.querySelectorAll('script[type="application/vnd.jupyter.widget-view+json"]');
+        for (let i=0; i!=tags.length; ++i) {
+            const viewtag = tags[i];
+            const widgetViewObject = JSON.parse(viewtag.innerHTML);
+            const { model_id } = widgetViewObject;
+            const model = models[model_id];
+            const widgetTag = document.createElement('div');
+            widgetTag.className = 'widget-subarea';
+            viewtag.parentElement.insertBefore(widgetTag, viewtag);
+            const view = await this.display_model(undefined, model, { el : widgetTag });
+            widgetTag.appendChild(view.node);
+        }
+    }
+
+    async loadClass(className, moduleName, moduleVersion) {
+        if (
+            moduleName === '@jupyter-widgets/base' ||
+            moduleName === '@jupyter-widgets/controls' ||
+            moduleName === '@jupyter-widgets/output'
+        ) {
+            return super.loadClass(className, moduleName, moduleVersion);
+        }
+        else {
+            return htmlManager.loadClass(className, moduleName, moduleVersion);

```

Comment:
I'd say, use the original loader idea. Since you already check those 3 modes, I think you could do:
```
...
 else 
      return this.loader(moduleName, moduleVersion)
```

Reply:
I am not sure I understand: for the base, controls and output modules I want to use the jupyterlab-manager loadClass, but for the rest of the modules I want to keep using the html-manager loadClass, which internally handles promise rejections and allows me to not replicate such code.

Reply:
I duplicated the code, I think in this case it's cleaner for the moment, so it's safer in case we want to refactor this.